File: hard/19_laptop_rentals/laptop_rentals.js
function laptopRentals(times) {
  if (times.length === 0) {
    return 0;
  }

  let usedLaptops = 0;
  const startTimes = times.map((interval) => interval[0]).sort((a, b) => a - b);
  const endTimes = times.map((interval) => interval[1]).sort((a, b) => a - b);

  let startIterator = 0;
  let endIterator = 0;

  while (startIterator < times.length) {
    if (startTimes[startIterator] >= endTimes[endIterator]) {
      usedLaptops -= 1;
      endIterator += 1;
    }

    usedLaptops += 1;
    startIterator += 1;
  }

  return usedLaptops;
}
console.log(laptopRentals([[0, 2], [1, 4], [4, 6], [0, 4], [7, 8], [9, 11], [3, 10]]));

module.exports = laptopRentals;

File: hard/19_laptop_rentals/laptop_rentals.go
package main

import (
	"fmt"
	"sort"
)

func laptopRentals(times [][]int) int {
	if len(times) == 0 {
		return 0
	}

	usedLaptops := 0
	startTimes := []int{}
	endTimes := []int{}

	for _, interval := range times {
		startTimes = append(startTimes, interval[0])
		endTimes = append(endTimes, interval[1])
	}

	sort.Ints(startTimes)
	sort.Ints(endTimes)

	startIterator := 0
	endIterator := 0

	for startIterator < len(times) {
		if startTimes[startIterator] >= endTimes[endIterator] {
			usedLaptops -= 1
			endIterator += 1
		}

		usedLaptops += 1
		startIterator += 1
	}

	return usedLaptops
}

func main() {
	fmt.Println(laptopRentals([][]int{{0, 2}, {1, 4}, {4, 6}, {0, 4}, {7, 8}, {9, 11}, {3, 10}}))
}
